pyannotate_runtime/collect_types.py

```diff
@@ -777,14 +852,25 @@ def _trace_dispatch(frame, event, arg):
                 resolved_types = prep_args(arg_info)
                 _task_queue.put(KeyAndTypes(function_key, resolved_types))
             elif event == 'return':
-                # This event is also triggered if a function raises an exception.
+                # This event is also triggered if a function yields or raises an exception.
                 # We can tell the difference by looking at the bytecode.
                 # (We don't get here for C functions so the bytecode always exists.)
-                # TODO: Also recognize YIELD_VALUE opcode.
                 last_opcode = code.co_code[frame.f_lasti]
-                if last_opcode != RETURN_VALUE_OPCODE:
-                    arg = NoReturnType()
-                _task_queue.put(KeyAndReturn(function_key, resolve_type(arg)))
+                if last_opcode == RETURN_VALUE_OPCODE:
+                    if code.co_flags & CO_GENERATOR:
+                        # Return from a generator.
+                        t = resolve_type(FakeIterator([]))
+                    else:
+                        t = resolve_type(arg)
+                elif last_opcode == YIELD_VALUE_OPCODE:
+                    # Yield from a generator.
+                    t = resolve_type(FakeIterator([arg]))
```

Comment:
If I understand this correctly, we are going to create `Iterator[x]` for every yielded value.
Would not it be better to accumulate the list of yielded values until we receive the `RETURN_VALUE_OPCODE` with `CO_GENERATOR` and then add 1 KeyAndReturn entry for them?

Reply:
Yeah, but the question is where does that list of yielded values get stored. The queue exists so the tracing code doesn't have to store anything (apart from sampling_counters and the queue).